Checks for Units in Component math in code generation

awickens opened the pull request:
Work towards creating verification for units in MathML strings as part of a `component`. 

Addresses #496.

Commit:
work towards initial math validation tests

Commit:
refactored first two unit tests, others are a work in progress

nickerso commented:
Just an initial observation, the `Validator` should be validating against the rules defined in the CellML specification. Checking dimensional consistency within equations is not part of validation (against the spec). These are not errors in the CellML model, they are errors in the mathematical model that is being encoded in CellML.

So I would expect this to be done by a something other than the `Validator` class or as a different method in the current `Validator`. And to probably really start using the extended `Issue` ideas from #368.

agarny commented:
> So I would expect this to be done by a something other than the `Validator` class or as a different method in the current `Validator`. And to probably really start using the extended `Issue` ideas from #368.

Agreed. Probably as part of checking whether a model is runnable (see issue #493), although I feel like that code should be extracted from our `Generator` class (see issue #499). Those issues are somewhat all related. So, we will need to be careful in which order they are tackled. I feel like issue #499 should be worked on first. Then this issue is really a subset of issue #493.

In the meantime, and as a prerequisite to those issues, there is indeed the `Issue` class that really ought to be implemented. So, that would probably by my priority at this stage.

Commit:
Moving unit tests to generator class in favour of having equation checking done within the AST present in the class

Commit:
Adding function declaration for AST units checking

Commit:
created helper function to instantiate AST for units checking

Commit:
Small progress made on validating math, still trying to figure out how best to use the ast to traverse variables

Commit:
Adding initial unit equivalence checking functions, subject to change

Commit:
Added internal node checking function, as it is easier than using lots of clunky if statements in the main function body

Commit:
function skeleton for main units analysis function altered to fit with new type function

Commit:
declaring shim function so implementation is hidden/private

Commit:
Minor cleaning up, removed unnecessary function declarations

Commit:
Cleaning up of skeleton of main recursive function

Commit:
Skeleton logic implemented for main recursive function processEquationUnits, most helper functions implemented

Commit:
Adding functions for adding/multiplying units in the units mapping

Commit:
Minor changes

Commit:
minor correction to function description

Commit:
Addresses invalid argument access when attempting to read leaf nodes

Commit:
Reimplemented AST internal node checking functions for better clarity

Commit:
Deleting old AST checkNodeType function

Commit:
Fixed seg fault exception when attempting to access variable for a non-variable cn node

Commit:
Fixed SEH exceptions, narrowed errors down to mappings only

Commit:
Solved mapping errors, now implementing more of the AST tree cases to reduce errors

Commit:
Temporary fix for invalid SEH exception when attempting to return error message for invalid unit comparison.

Commit:
Recovering changes

Commit:
Changed processModel so processEquationUnits is in it's own separate loop, to prevent error double ups

Commit:
Adding a type to string mapping to output operation where error was found

Commit:
Adding/fixing units implementation in the AST

Commit:
Fixes in main AST checking function to take advantage of new mUnits attribute

Commit:
Added condition in case we want to declare a negative parameter

awickens committed:
Merge branch 'develop' into math_validation

Commit:
Merge branch 'math_validation' of https://github.com/awickens/libcellml into math_validation

Commit:
minor cleanup before reimplementing power/root checks in recursive function

Commit:
Fixed bug in cellml_mappings_and_encapsulations model.cellml file

Commit:
Added check for standard unit names, rearranged updateBaseUnitCount function to allow for units not defined in model to be checked first

Commit:
Fixed standardRef bug which threw invalid map exception in cellmlMappingsAndEncapsulations

Commit:
Changed incorrect SI unit for Farad in utilities.h: should be second^4.0, not second^-4.0

Commit:
Changed incorrect definition of coulomb in utilities.h, should be ampere^1.0, not ampere^-1.0

Commit:
Adding fix for generator test, so there is no invalid stod exception thrown

Commit:
Small fix to ensure correct mapping returned on operations which only have one child

Commit:
minor cleaning up

Commit:
Getting rid of commented code

Commit:
Merge branch 'develop' of https://github.com/cellml/libcellml into math_validation

Commit:
Minor fixes to ensure Mac and Ubuntu compile

Commit:
Cleaning up main recursive function, hints moved to another function

Commit:
Changing string declarations in event of no model to empty string, not nullptr

agarny commented on src/generator.cpp:
```diff
@@ -1682,8 +1682,8 @@ UnitsMap processEquationUnitsAst(const GeneratorEquationAstPtr &ast, UnitsMap un
                     ModelPtr model = (component != nullptr) ? owningModel(component) : nullptr;
-                    std::string compName = (component != nullptr) ? component->name() : nullptr;
-                    std::string modelName = (model != nullptr) ? model->name() : nullptr;
+                    std::string compName = (component != nullptr) ? component->name() : " ";
```
Can't help but wonder why `" "` rather than `""`? (Keeping in mind that I have no idea what the code is supposed to be doing... :))

awickens replied:
No reason at all, actually. I'm just committing to make sure unit tests pass but coverage certainly won't - I'll change it once I'm satisfied with the test I have (which is a work in progress, I got rid of the old ones).

Commit:
Formatting code

Commit:
adding test as work towards defining behaviour and getting full coverage

Commit:
Adding test, slightly changingmain recursive function logic

Commit:
Getting rid of redundant line for coverage

Commit:
Getting rid of redundancy in updateBaseUnitCount, left in as something to potentially discuss whether necessary or not

Commit:
format code

awickens commented:
Currently at the stage where the units checking works - still need to check for multipliers, though -  but it would be nice if anyone shared their thoughts on the error messages. I was thinking of putting the actual mathematical expression into the error message, rather than just the operation and the component it is in (although it was sufficient when debugging some of the larger models in `test_generator`).

kerimoyle commented:
Hey @awickens It might be worth holding off on the error messages for a bit - we're trying to figure out some "rules" for them at the moment (see #514), so any feedback on them now would be probably subject to change!

awickens commented:
Right, makes sense. I'll hold off on making any changes to the messages for the moment.

agarny commented:
Agreed with @kerimoyle, don't worry too much about the error message for now.

Commit:
Adding test for multipliers in model to generator resources

Commit:
Adding associated test in test_generator

Commit:
Adding in processEquationMultiplierAst function

Commit:
Adding first comparison case for multipliers

Commit:
Adding derivative case for multipliers

Commit:
Adding bvar case for multipliers

Commit:
Adding multiply/divide case for multipliers

Commit:
Adding power/root case for multipliers

Commit:
Added initial trig case

Commit:
Fix for bottom variable NaN bug in code

Commit:
Reduced failed test cases to six, issues still in multiplier comparisons further up the AST

kerimoyle commented on src/generator.cpp:
```diff
@@ -1515,10 +1515,14 @@ void updateBaseUnitCount(const ModelPtr &model,
-        } else if (unitMap.find(uName) == unitMap.end()) {
+        } 
+        
+        // Leaving this as a comment for the moment: This would only be necessary if we had a base unit which was *not* in the standard units list - uncertain if this will ever occur within a formal model.
```
Yes, it will still be needed.  Users should be able to define their own custom base units and build others from them.  Most obvious units that don't use our current base variables are for cost/dollars/francs/whatever or pH.

Commit:
Merge branch 'develop' of https://github.com/cellml/libcellml into math_validation

Commit:
Small fix for multiplier calculations in bottom variable AST

Commit:
Fix to updateBaseUnit to pass cellmlMappingsAndEncapsulations

Commit:
Minor changes, five tests still failing

Commit:
Changed getPower to remove extra case

Commit:
Fixed error in updateBaseUnitCount, all tests now pass

Commit:
Cleaning up code, fixed issues so tests pass

Commit:
Minor fixes for compile

agarny commented on src/generator.cpp:
```diff
@@ -1868,7 +1868,7 @@ double processEquationMultiplierAst(const GeneratorEquationAstPtr &ast, std::vec
             // The only time we check multiplier mismatch is in a comparision operation.
             if (isDirectComparisonOperator(ast)) {
-                if (leftMult != rightMult && ast->mLeft != nullptr && ast->mRight != nullptr) {
+                if ((leftMult > rightMult || leftMult < rightMult) && ast->mLeft != nullptr && ast->mRight != nullptr) {
```
Wow, this is *ugly*! Please use [`areEqual()`](https://github.com/cellml/libcellml/blob/4f19a5c917c60f7f345ca11d9f7649d33d8b3bac/src/utilities.cpp#L163-L166) instead.

Commit:
added areEqual() to multiplier check

Commit:
multiplier fix to ensure compile on macOS

Commit:
Changed function to pass in multiplier as a parameter

Commit:
Adding extra test cases to satisfy coverage

Commit:
Coverage fix attempt 2

awickens commented:
I know this can't really be merged in until #521 has been approved (so the errors can be changed to warnings instead), but any review of the code would be appreciated.

kerimoyle committed:
Merge branch 'develop' into math_validation

kerimoyle commented on src/utilities.h:
```diff
@@ -44,9 +44,9 @@ const std::map<std::string, std::map<std::string, double>> standardUnitsList = {
+    {"coulomb", {{"ampere", 1.0}, {"second", 1.0}}},
     {"dimensionless", {{"dimensionless", 1.0}}},
-    {"farad", {{"ampere", 2.0}, {"kilogram", -1.0}, {"metre", -2.0}, {"second", -4.0}}},
+    {"farad", {{"ampere", 2.0}, {"kilogram", -1.0}, {"metre", -2.0}, {"second", 4.0}}},
```
Well caught! This was my bad, taken from a typo in the spec.

kerimoyle commented on tests/resources/generator/validate_units_in_model.cellml:
```diff
@@ -0,0 +1,232 @@
+                </apply>
+            </apply>
+        </math>
+        <!-- Not allowed - 2 is not dimensionless-->
```
Do you also have a test when the power is a variable (ie: <ci> vs <cn>), to test that it's dimensionless?

awickens replied:
I will have to double check, but I think in some of the other models used as unit tests which @agarny put in (perhaps in the coverage test as well) there are instances of variables used as powers.

agarny replied:
Indeed, the coverage test (at least) has instances of variables used as powers ([here](https://github.com/cellml/libcellml/blob/7e991e0f69d2512576659d1c41ef068ae95f7c27/tests/resources/generator/coverage/model.cellml#L396-L412) for example).

kerimoyle replied:
Are we even allowed to have dimensional stuff as an exponent?  How on earth do the units handle that?

agarny replied:
As far as I can remember, only a `dimensionless` exponent makes sense. This is the reason `n` in `m^n` is `dimensionless`.

awickens replied:
Yep, @kerimoyle it doesn't make sense to have dimensional stuff as exponents (obviously), but to the best of my recollection there is nothing preventing someone from writing a CellML file intentionally with exponents which are not dimensionless, hence why it is still good to check if it is in case someone's messed up the CellML or incorrectly defined a variable with units.

kerimoyle commented on src/generator.cpp:
```diff
@@ -1285,6 +1313,649 @@ void Generator::GeneratorImpl::processEquationAst(const GeneratorEquationAstPtr
+    return type == libcellml::GeneratorEquationAst::Type::DIFF;
+}
+
+bool isBottomVariableOperator(const GeneratorEquationAstPtr &ast)
```
I think that the "b" in "bvar" stands for "bound"? Pretty sure it's not bottom :)

awickens replied:
Yep, will change tomorrow.

kerimoyle commented on src/generator.cpp:
```diff
@@ -1285,6 +1313,649 @@ void Generator::GeneratorImpl::processEquationAst(const GeneratorEquationAstPtr
+    }
+}
+
+void updateBaseMultiplier(const ModelPtr &model,
```
See PR #526 ... there was a bug somewhere around this area, just need to check that yours is ok?

awickens replied:
Yes, double checked now and I have the correct format (with the ```standardMultiplierList.at(uName)```) which was missed out in ```units.cpp```.

kerimoyle commented on src/generator.cpp:
```diff
@@ -1285,6 +1313,649 @@ void Generator::GeneratorImpl::processEquationAst(const GeneratorEquationAstPtr
+            return unitMap;
+        }
+
+        // We know if we have reached an internal vertex that we have a mathematical operation as it's type.
```
its not it's

kerimoyle commented on src/generator.cpp:
```diff
@@ -1285,6 +1313,649 @@ void Generator::GeneratorImpl::processEquationAst(const GeneratorEquationAstPtr
+                }
+
+                // If we have a square root operation
+                if (power == 0.0 && ast->mRight == nullptr) {
```
I don't quite follow this bit?

awickens replied:
There are instances in the AST where a root number is not specified as a child (i.e, the root node will only have a variable child and no second child giving a root number - it happens a couple of times in some of the bigger models in the unit tests for the `validator`). So, if there is no child (a power cannot be found, so 0.0 is returned), the assumption is that the intended operation is a square root operation.

nickerso replied:
MathML defines a default square-root: https://www.w3.org/TR/REC-MathML/chap4_4.html#sec4.4.3.11

kerimoyle replied:
I'm not sure we should be making any assumptions anywhere?  Isn't that the point of an unambiguous model description? Rather return an error ...

awickens replied:
I think a few of the models used for testing (not the ones I added in) would spontaneously combust if I was to just return an error. Perhaps when the AST gets moved out/exposed then some discussion can be made as to whether default roots like ```sqrt``` can be made into their own type of internal node.

nickerso replied:
the definition of the root operator is very clear in this regard. It is not an assumption. A root with two arguments has the degree (number) and the thing it is applied to; a root with a single argument has degree 2. (sorry if I'm missing something, trying to be helpful without reading everything ;-)

kerimoyle commented on src/utilities.h:
```diff
@@ -44,9 +44,9 @@ const std::map<std::string, std::map<std::string, double>> standardUnitsList = {
     {"becquerel", {{"second", -1.0}}},
     {"candela", {{"candela", 1.0}}},
-    {"coulomb", {{"ampere", -1.0}, {"second", 1.0}}},
+    {"coulomb", {{"ampere", 1.0}, {"second", 1.0}}},
```
And this one was just my mistake

kerimoyle commented:
HI @awickens!  I've had a quick peek and it's looking good (and I'm so glad you caught my units typos!)  I wouldn't worry about waiting for creating `Error`s or waiting for the `Issue` class - just make the errors here as best you can, but realise it could change once we're certain about how to finesse the messages.  It is useful to understand what kinds of problems will be reported though, so the error messages are a bit chicken and egg.  

I did wonder about whether the generator checks of maths should be done independently of the `Generator` class though?  So that people can check for errors in their maths/model formulation without necessarily needing to generate code to do it?  @agarny this is what you and I talked about a few weeks back? Thoughts?

agarny commented:
> I did wonder about whether the generator checks of maths should be done independently of the `Generator` class though? So that people can check for errors in their maths/model formulation without necessarily needing to generate code to do it? @agarny this is what you and I talked about a few weeks back? Thoughts?

Indeed, this should be done as part of issues #493/#499.

Commit:
Added functions to better express where error messages have occurred

Commit:
Adding error functions into main error messages

Commit:
Merge branch 'math_validation' of https://github.com/awickens/libcellml into math_validation

awickens commented:
Ready for review.

kerimoyle committed:
Merge branch 'develop' into math_validation

awickens committed:
Changing function name to bound variable operator, not bottom

awickens committed:
Merge branch 'math_validation' of https://github.com/awickens/libcellml into math_validation

Merging develop into math_validation

awickens committed:
Merge branch 'develop' into math_validation

hsorby committed:
Merge branch 'develop' into math_validation

nickerso committed:
Merge branch 'develop' into math_validation

agarny commented:
Just wondering... is this PR ready? If so, I might take over it and update it with our current `develop` branch. @awickens, @hsorby, @kerimoyle and @nickerso?

awickens commented:
@agarny coming out of the woodwork after a couple months, this was probably ready in March but I wouldn't be certain now, especially since there seems to be some code reshuffling going on where units checking is done elsewhere (looking at the Analyser class being implemented).

agarny commented:
> @agarny coming out of the woodwork after a couple months, this was probably ready in March but I wouldn't be certain now, especially since there seems to be some code reshuffling going on where units checking is done elsewhere (looking at the Analyser class being implemented).

Thanks @awickens, it's true that a lot of reshuffling of code has happened (and some more is on its way). So, just wondering, is that something on which you will be keen to work (once PRs #631 and #650) have been merged in, or would you rather leave it to me? (No pressure whatsoever! 🙂)

awickens commented:
> 
> Thanks @awickens, it's true that a lot of reshuffling of code has happened (and some more is on its way). So, just wondering, is that something on which you will be keen to work (once PRs #631 and #650) have been merged in, or would you rather leave it to me? (No pressure whatsoever! 🙂)

It depends on when you'd like it to be done (although I'd like to finish what I started!). I'm out through the South Island until the 27th, so I wouldn't be able to touch it for a couple of weeks until I get back up to Auckland. If PRs #631 and #650 get merged in within the next couple of days I would leave it up to you to get it done :)

agarny commented:
> It depends on when you'd like it to be done (although I'd like to finish what I started!). I'm out through the South Island until the 27th, so I wouldn't be able to touch it for a couple of weeks until I get back up to Auckland. If PRs #631 and #650 get merged in within the next couple of days I would leave it up to you to get it done :)

There is no rush, I was merely interested to know whether you were still keen to work on this issue since I agree that you should finish what you started. PR #631 is ready to go, but not PR #650 (it will, now, likely be ready sometime next week), so we can definitely wait until you are back.

awickens committed:
Fixing merge conflicts

awickens committed:
Merge branch 'develop' into math_validation

awickens commented:
Okay @agarny, I've finally gotten around to this (about time, too). I just thought I'd run my plan by you to make sure I don't end up moving code over to the incorrect thing and having to revert it all back.

In my head, I envisioned moving this stuff from the ```generator``` class over into the ```analyser``` class, since all the AST exposure stuff has been moved there. This also includes moving the tests over (from memory I made a couple CellML files to check behavior). I suspect the error messages will have changed significantly from when I implemented them as well, so those will likely have to be changed as well. I was wondering if there was anything else in particular which needed to be addressed as well?

agarny commented:
> Okay @agarny, I've finally gotten around to this (about time, too).

Very nice to see some progress on this. Welcome back! :)

> I just thought I'd run my plan by you to make sure I don't end up moving code over to the incorrect thing and having to revert it all back.
>
> In my head, I envisioned moving this stuff from the `generator` class over into the `analyser` class, since all the AST exposure stuff has been moved there. This also includes moving the tests over (from memory I made a couple CellML files to check behavior). I suspect the error messages will have changed significantly from when I implemented them as well, so those will likely have to be changed as well. I was wondering if there was anything else in particular which needed to be addressed as well?

Yes, to move your work to the analyser is definitely the way forward. About something else that might need to be addressed, I will have to get back to you on this. Like you, it has been a long time since I have looked into this issue. Still, it should be safe to start moving your work to the analyser. (PRs #650 and #763 should be merged on Monday, but hopefully it shouldn't interfere too much with your work.)

awickens committed:
Cleaning up generator source, old functions were left in there after merge.

awickens committed:
Removing equation tree checking tests and testing files from the test/generator unit tests

awickens committed:
Merge branch 'develop' into math_validation

hsorby committed:
Remove commented out code.

hsorby commented:
Is anyone else seeing just the one file changed here?

awickens commented:
@hsorby Yeah, I had taken out all the stuff from the ```validator``` (with associated tests) awhile back, in preparation to move it over - unfortunately commitments to what I am doing over this summer have gotten in the way (plus Visual Studio refusing to build properly). I can start moving stuff back in once I sort out what's wrong with Visual Studio.

hsorby commented:
Hi @awickens, we are looking to move this forward because we want to use it for the building blocks of handling DAEs.  We will start working on this from 02/02/21.  Please don't stress if you cannot do what you were planning by then, we are happy to move this forward from the work you have done.  We are of course still happy to have your contributions at any time.  This is just a note to let you know so we don't double up on work.

awickens committed:
Adding validator code as comments to direct implementation in analyser class

awickens committed:
Merge branch 'math_validation' of https://github.com/awickens/libcellml into math_validation

awickens committed:
Merge branch 'develop' into math_validation

awickens commented:
Hi @hsorby, I've put all the changes I originally made in ```validator.cpp``` as comments in the ```analyser.cpp``` file (as well as the tests I made all that time ago in ```test_analyser.cpp```) - hopefully this helps as this gets moved forwards from today.

agarny commented:
Thanks a lot @awickens.

agarny commented:
@hsorby, I will have a look at it tomorrow, if you don't mind.

agarny committed:
Make ClangFormat happy.

agarny committed:
Merge branch 'develop' into pr/497.

agarny committed:
Analyser tests: renamed some test CellML files and updated our unit validation related tests.

agarny committed:
Analyser tests: initial clean up of our unit validation related tests.

agarny committed:
Some minor cleaning up.

agarny committed:
Utilities: added a way to check whether a unit is dimensionless.

agarny committed:
Analyser: added the skeleton for analyseEquationUnits().

agarny committed:
Analyser: keep track of the units of a CI/CN element.

agarny committed:
Analyser: re-added units checking.

Still a lot of cleaning up to do and, most likely, need to implement a more comprehensive testing strategy.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser tests: further clean up of our unit validation related tests.

agarny committed:
Some minor cleaning up.

agarny committed:
Make Linux happy.

agarny committed:
Make macOS happy.

agarny committed:
Make macOS happy.

agarny committed:
Didn't make macOS happy so reverting.

agarny committed:
Make macOS happy.

agarny committed:
Analyser: prevent crash on Windows and Linux.

Temporary solution since I don't think we should have this expression() method.

agarny committed:
Analyser: prevent crash on Windows and Linux.

agarny committed:
Utilities: removed isDimensionlessUnit().

Indeed, we don't use it and therefore can't have coverage for it in our tests.

agarny committed:
Analyser: commented out some code that is not (currently?) used.

agarny committed:
Analyser: commented out some code that is not (currently?) needed.

agarny committed:
Analyser: commented out some code that is not (currently?) needed.

agarny committed:
Make ClangFormat happy.

agarny committed:
Utilities: removed isDimensionlessUnit().

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: make use of the isDimensionlessUnit() method.

agarny committed:
Analyser: cleaned up our analyseEquationUnits() method.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: changed the way we keep track of the unit associated with an AST leaf.

Indeed, we previously kept track of it in the AST itself, which was not only the wrong place (AST users don't need that piece of information) and a waste of memory (only CN/CI leaves have a unit).

agarny committed:
Replaced calls to std::map::emplace() with std::map::operator[]().

std::map::emplace() is not needed as such in those cases (at least, our tests don't fail). Also, makes our code easier to read and maybe very slightly faster.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: keep track of all CN units.

Even those that are dimensionless. Indeed, we (automatically) do it for CI elements, so we may as well do it for CN elements.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: only keep track of non-dimensionless CI/CN units.

Kind of reverted commit 54486f108. Indeed, no point in wasting memory and time by keeping track of everything. Instead, we should only keep track of what is relevant.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

hsorby commented on src/analyser.cpp:
```diff
@@ -1197,9 +1197,13 @@ void Analyser::AnalyserImpl::updateUnitsMap(const ModelPtr &model,
+            if (unitsMap.find(unitsName) == unitsMap.end()) {
+                unitsMap[unitsName] = unitsExponent;
+            } else {
+                auto iter = unitsMap.find(unitsName);
```
Not sure we need to do the find twice here.

agarny replied:
Correct, I was just testing something. This will be corrected in a minute, although I will likely push it later with more upcoming changes.

agarny replied:
Still needed to push things, so that I can test locally on Windows/Linux. So, yes, addressed in commit 12b600a65.

agarny committed:
Some minor cleaning up.

agarny committed:
Make MSVC happy.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: use the generator for unit-related messages that require mentioning an expression/equation.

agarny committed:
Analyser: use the generator for unit-related messages that require mentioning an expression/equation.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: fixed a problem with units checking for CN elements with a standard unit.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Utilities: don't need isDimensionlessUnit() anymore.

Also removed isDimensionlessUnitName() and test directly against "dimensionless".

agarny committed:
Some minor cleaning up.

agarny committed:
Analysis: simplified the retrieval of a component name for an AST.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

Also fixed a small problem in componentName(). We were checking the left child twice!), hence our coverage test was wrong (!!).

agarny committed:
Analyser: renamed one of our test files.

Typo...!

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Some minor cleaning up.

Fix commit 295d9d7b9.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

These are the messages we would normally expect to see for a piecewise statement.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: allow for multi-level units checks (needed for piecewise statements).

agarny committed:
Analyser: make MSVC happy.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: properly handle piecewise elements when units checking.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: some minor cleaning up.

That coverage test is now handled in our piecewise test.

agarny committed:
Analyser: added some units-based tests for relational operators.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: added some units-based tests.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: added some units-based tests for `plus`, `minus`, `times` and `divide`.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
Generator: some minor improvement.

agarny committed:
Analyser: some minor cleaning up.

Some of our units tests have a name which is a C++ keyword, so use `_xxx_` since otherwise `not` would result in some strange Clang formatting.

agarny committed:
Analyser: some minor cleaning up.

hsorby reviewed (commented):
You can't start identifiers with underscore in C++ that is reserved.

agarny committed:
Analyser: some minor cleaning up.

Fixes commit 4632110bf due to identifiers not starting with an underscore in C++.

agarny commented:
> You can't start identifiers with underscore in C++ that is reserved.

Argh, why does Clang accept it?! Anyway, fixed in commit b591d08c5.

agarny committed:
Analyser: simplified and fixed units checking for the root function.

agarny committed:
MathML: fixed some of our root-related tests.

agarny committed:
Analyser: added some units-based tests for `power` and `root`.

agarny committed:
MathML: fixed some of our root-related tests.

agarny committed:
Generator: fixed a problem with the root function/operator.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Generator: fixed a problem with the root function/operator.

agarny committed:
Analyser: added some units-based tests for `abs`.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: properly handle piecewise elements when units checking.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: commented what we expect when it comes to units and relational/logical operators.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: commented what we expect when it comes to units and MathML elements.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Merge branch 'develop' into pr/497.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny committed:
Analyser: improved our units-based messages.

agarny commented:
This is replaced with PR #818, following the renaming of our `develop` branch to `main` (and `master` to `release`).
